extensions/account-registration.md

```diff
@@ -0,0 +1,188 @@
+for future organic expansion.
+
+### Capability
+
+This specification adds the `draft/register` capability, whose presence
+signifies that the server accepts the `REGISTER` command.
+
+The capability has an optional value, a comma-separated list of key-value
+pairs; the format is intended to follow the precedent set by
+[the multiline draft][multiline].
+
+The defined keys are:
+
+ * `before-connect` - if present, indicates the server supports early
+   registration, so it MUST NOT use
+   `FAIL REGISTER COMPLETE_CONNECTION_REQUIRED`
+ * `email-required` - if present, registrations require a valid email address
+   to process
+
+Clients MUST ignore any value assigned to these keys, and MUST ignore
+any unknown key.
+
+Software implementing this work-in-progress specification MUST NOT use
+the unprefixed `register` or `account-registration` capability names.
+Instead, implementations SHOULD use the `draft/register`
+capability name to be interoperable with other software implementing
+a compatible work-in-progress version.
+
+### Commands
+
+    REGISTER {<email> | "*"} <password>
+    
+The `REGISTER` command informs the server of a request to register
+an account named for the current nick of the requestor.
+
+The `REGISTER` command MAY be sent at any point during the connection
+that the client has a valid nickname.
+If the client sends `REGISTER` before completing connection registration,
+and receives a `FAIL REGISTER COMPLETE_CONNECTION_REQUIRED`, it SHOULD make
+a second attempt after it receives the welcome message.
+
+    VERIFY <account> <code>
+    
+The `VERIFY` command completes a registration that required verification
+(eg. via email or CAPTCHA).
+
+### Responses
+
+    REGISTER SUCCESS <account> <message>
+    VERIFY SUCCESS <account> <message>
+    
+Sent by the server when the `REGISTER` or `VERIFY` fully succeeded, and no
+further action is needed.
+Afterward, the client is assumed to be authenticated to the `<account>`
+as if it used SASL.
+
+    REGISTER VERIFICATION_REQUIRED <account> <message>
+    
+Sent by the server as a response to `REGISTER` when further action is required
+to complete registration of the `<account>`, as explain in the human-readable
+`<message>`.
+
+    FAIL REGISTER ACCOUNT_EXISTS <account> <message>
+
+Sent by the server as a response to `REGISTER` when the client tried to register
+while using a nick that is not available as a new account's name
+(for example, because someone already registered it).
+
+The server MUST NOT send this response before the client sent a `NICK` command.
+
+The client MAY try registering again later.
+Clients should not automatically retry immediately without changing
+their nickname or waiting.
+
+    FAIL REGISTER NEED_NICK * <message>
+
+Sent by the server if the client sends `REGISTER` before sending a NICK command.
+This is only possible in setups that advertise the `before-connect` key.
+
+    FAIL REGISTER ALREADY_REGISTERED <account> <message>
```

Comment:
The `ALREADY_REGISTERED` fail code is sent if a client is already *authenticated*.  Should this be `ALREADY_AUTHENTICATED`?

Reply:
makes sense